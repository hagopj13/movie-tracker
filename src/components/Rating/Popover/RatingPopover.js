// @flow
import React, { useState } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Popover from '@material-ui/core/Popover';
import Box from '@material-ui/core/Box';
import Rating from '@material-ui/lab/Rating';

type Props = {
  isOpen: boolean,
  anchorEl: HTMLElement | null,
  initialValue: number,
  onClose: () => void,
  onChange: (newValue: number | null) => void,
};

const useStyles = makeStyles((theme) => ({
  content: {
    display: 'flex',
    alignItems: 'center',
    padding: theme.spacing(1.5),
    width: 165,
  },
}));

const RatingPopover = (props: Props) => {
  const { isOpen, anchorEl, initialValue, onClose, onChange } = props;

  const [value, setValue] = useState(initialValue);
  const [hover, setHover] = useState(-1);

  const classes = useStyles();

  const handleChange = (event, newValue: number | null) => {
    setValue(newValue);
    onChange(newValue);
  };

  const handleChangeActive = (event, newHover: number) => {
    setHover(newHover);
  };

  return (
    <Popover
      open={isOpen}
      anchorEl={anchorEl}
      onClose={onClose}
      anchorOrigin={{ vertical: 'bottom', horizontal: 'center' }}
      transformOrigin={{ vertical: -10, horizontal: 'center' }}
    >
      <div className={classes.content}>
        <Rating
          name="rating"
          value={value}
          precision={0.5}
          onChange={handleChange}
          onChangeActive={handleChangeActive}
        />
        <Box component="span" ml={1} fontSize={15}>
          {hover !== -1 ? hover : value}
        </Box>
      </div>
    </Popover>
  );
};

export default RatingPopover;
