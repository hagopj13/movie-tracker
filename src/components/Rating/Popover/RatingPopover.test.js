import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';
import Popover from '@material-ui/core/Popover';
import Rating from '@material-ui/lab/Rating';

import RatingPopover from './RatingPopover';

describe('RatingPopover component', () => {
  let shallow;
  let isOpen;
  let initialValue;
  let wrapper;

  const mockClose = jest.fn();
  const mockChange = jest.fn();

  beforeAll(() => {
    shallow = createShallow();
  });

  beforeEach(() => {
    jest.clearAllMocks();
    isOpen = true;
    initialValue = 5;
    const mockProps = {
      isOpen,
      initialValue,
      onClose: mockClose,
      onChange: mockChange,
    };
    wrapper = shallow(<RatingPopover {...mockProps} />);
  });

  it('should render the RatingPopover component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should pass isOpen prop to Popover', () => {
    expect(wrapper.find(Popover).prop('open')).toBe(isOpen);
  });

  it('should set the initial value of Rating', () => {
    expect(wrapper.find(Rating).prop('value')).toBe(initialValue);
  });

  it('should change the value and call onChange when rating changes', () => {
    const newValue = 6;
    wrapper.find(Rating).prop('onChange')(null, newValue);
    expect(wrapper.find(Rating).prop('value')).toBe(newValue);
    expect(mockChange).toHaveBeenCalledTimes(1);
    expect(mockChange).toHaveBeenCalledWith(newValue);
  });

  it('should call onClose prop when Popover is closed', () => {
    wrapper.find(Popover).prop('onClose')();
    expect(mockClose).toHaveBeenCalledTimes(1);
  });
});
