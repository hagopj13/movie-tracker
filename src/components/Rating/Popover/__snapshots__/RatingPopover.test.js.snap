// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RatingPopover component should render the RatingPopover component correctly 1`] = `
<WithStyles(ForwardRef(Popover))
  anchorOrigin={
    Object {
      "horizontal": "center",
      "vertical": "bottom",
    }
  }
  onClose={[MockFunction]}
  open={true}
  transformOrigin={
    Object {
      "horizontal": "center",
      "vertical": -10,
    }
  }
>
  <div
    className="makeStyles-content-1"
  >
    <WithStyles(ForwardRef(Rating))
      name="rating"
      onChange={[Function]}
      onChangeActive={[Function]}
      precision={0.5}
      value={5}
    />
    <Styled(MuiBox)
      component="span"
      fontSize={15}
      ml={1}
    >
      5
    </Styled(MuiBox)>
  </div>
</WithStyles(ForwardRef(Popover))>
`;
